flink-addons/flink-language-binding/src/main/java/org/apache/flink/languagebinding/api/java/common/OperationInfo.java

```diff
@@ -0,0 +1,48 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.flink.languagebinding.api.java.common;
+
+/**
+ * Container for all generic information related to operations. This class contains the absolute minimum fields that are
+ * required for all operations. This class should be extended to contain any additional fields required on a
+ * per-language basis.
+ */
+public abstract class OperationInfo {
+	public int parentID; //DataSet that an operation is applied on
+	public int otherID; //secondary DataSet
+	public int setID; //ID for new DataSet
+	public int[] keys1; //grouping keys
+	public int[] keys2; //grouping keys
+	public int[] projectionKeys1; //projection keys
+	public int[] projectionKeys2; //projection keys
+	public Object types; //an object that is of the same type as the output type
```

Comment:
Mh. I don't think that the `getProducedType()` is called after the local "pre-flight" phase has finished.
So you can safely put the TypeInformation object into a `transient` field, so that the system doesn't try to serialize it.